service/serializer/src/main/java/org/jolokia/service/serializer/GSonSerializer.java

```diff
@@ -0,0 +1,201 @@
+                json = JsonPath.parse(json, jsonPathconfiguration).read(path).toString();
+            }
+            return wrapString(json);
+        } catch (PathNotFoundException e) {
+            getJolokiaContext().error(String.format("Unexpected error when serializing %s", pValue.getClass().getSimpleName()), e);
+            return null;
+        } catch (InvalidJsonException e) {
+            getJolokiaContext().error(String.format("Unexpected error when serializing %s", pValue.getClass().getSimpleName()), e);
+            return null;
+        } catch (JsonIOException e) {
+            getJolokiaContext().error(String.format("Unexpected error when serializing %s", pValue.getClass().getSimpleName()), e);
+            return null;
+        }
+    }
+
+    public Object deserialize(String pExpectedClassName, Object pValue) {
+        if (pValue == null) {
+            return null;
+        }
+        Class<?> expectedClass = ClassUtil.classForName(pExpectedClassName);
+        if (expectedClass == null) {
+            return null;
+        }
+        if (pValue instanceof String) {
+            if (((String) pValue).isEmpty() && expectedClass == String.class) {
+                return pValue;
+            }
+        }
+        try {
+            Gson gson = getGson();
+            return gson.fromJson(toJSON(pValue).toJSONString(), expectedClass);
+        } catch (JsonSyntaxException e) {
+            getJolokiaContext().error(String.format("Unexpected error when deserializing %s", pValue), e);
+            return null;
+        }
+    }
+
+    private Gson getGson() {
+        return new GsonBuilder()
+                .serializeNulls()
+                .serializeSpecialFloatingPointValues()
+                .create();
+    }
+
+    public WriteRequestValues setInnerValue(Object pOuterObject, Object pNewValue, List<String> pPathParts) throws AttributeNotFoundException, IllegalAccessException, InvocationTargetException {
+        if (pOuterObject == null) {
+            return null;
+        }
+        String originalJson = ((JSONAware) serialize(pOuterObject, Collections.<String>emptyList(), null)).toJSONString();
+        JsonElement newValue = (JsonElement) deserialize(JsonElement.class.getName(), pNewValue);
+        JSONAware originalValue = (JSONAware) serialize(pOuterObject, pPathParts, null);
+
+        String updatedJson;
+        if (originalValue == null) {
+            String lastPathElement = pPathParts.remove(pPathParts.size() - 1);
+            String path = buildPath(pPathParts);
+            updatedJson = JsonPath.parse(originalJson, jsonPathconfiguration).put(path, lastPathElement, newValue).jsonString();
+        } else {
+            String path = buildPath(pPathParts);
+            updatedJson = JsonPath.parse(originalJson, jsonPathconfiguration).set(path, newValue).jsonString();
+        }
+        return new WriteRequestValues(deserialize(pOuterObject.getClass().getName(), updatedJson), originalValue);
+    }
```

Comment:
I tried to get rid of the `setInnerValue()` method in the interface. Unortunately its quite involved since the (recursive) internal JSON serialization is highly interwoven with the data extraction part needed to extract the object on which to set the value.

For the moment I can see two solutions:

* Allow for an 'UnsupportedOperation' exception when a serializer doesn't support setting an inner value with a write operation. I don't know how often this functionality is used in the wild, but I have the feeling its a too small feature to justify the amount of work needed to restructure this whole module. An implementation can still decide whether it wants to support this, but need todo its own reflection to obtain an inner Java object.

* Move the extractor part from the Jolokia serializer in a core package, so that the reflection code used for setting inner values can be used directly and one could remove the `setInnerValue()` from the interface. The drawback here is, that moving this part into core, 90% of the serialization code would move, too, defeating the modularisation and replaceability of the serializer component. 

Any thought, what would be the better approach ?